caput/mpiarray.py

```diff
@@ -1069,6 +1153,203 @@ def _partition_io(self, skip=False, threshold=1.99):
+                    dist_axis = array.axis
+                else:
+                    if dist_axis != array.axis:
+                        raise AxisException(
+                            "The distributed axis for all MPIArrays in an expression should be the same"
+                        )
+
+                args.append(array.local_array)
+            else:
+                args.append(array)
+
+        return (args, dist_axis)
+
+    # pylint: disable=inconsistent-return-statements
+    # array_ufunc is a special general function
+    # which facilitates the use of a diverse set of ufuncs
+    # some which return nothing, and some which return something
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        """Handles ufunc operations for MPIArray.
+
+        In NumPy, ufuncs are the various fundamental operations applied to
+        ndarrays in an element-by-element fashion, such as add() and divide().
+        https://numpy.org/doc/stable/reference/ufuncs.html
+
+        ndarray has lots of built-in ufuncs. In order to use them, the MPIArrays
+        need to be converted into ndarrays, otherwise NumPy reports a
+        NotImplemented error.
+
+        The distributed axis for all input MPIArrays, is expected to be the same.
+        Operations across the distributed axis, will not be permitted.
+
+        The new array will either be distributed over that axis, or possibly
+        one axis down for `reduce` methods.
+
+        For operations that normally return a scalar, the scalars will be
+        wrapped into a 1D array, distributed across axis 0.
+
+        Parameters
+        ----------
+        ufunc: <function>
+            ufunc object that was called
+        method: str
+            indicates which ufunc method was called.
+            one of "__call__", "reduce", "reduceat", "accumulate", "outer", "inner"
+        inputs: tuple
+            tuple of the input arguments to the ufunc.
+            At least one of the inputs is an MPIArray.
+        kwargs: dict
+            dictionary containing the optional input arguments of the ufunc.
+            Important kwargs considered here are 'out' and 'axis'.
+        """
+        # pylint: disable=no-member
+        # known problem with super().__array_ufunc__
+
+        args = []
+
+        # convert all local arrays into ndarrays
+        args, dist_axis = MPIArray._mpi_to_ndarray(inputs)
+
+        if "axis" in kwargs and (kwargs["axis"] == dist_axis):
+            raise AxisException(
+                f"operations along the distributed axis (in this case, {dist_axis}) are not allowed."
+            )
+
+        # 'out' kwargs contain arrays that the ufunc places the results into
+        # this views the local part of the output arrays into an ndarray
+        # that the ufunc knows how to work with
+        outputs = kwargs.get("out", None)
+        if outputs:
+            out_args, _ = MPIArray._mpi_to_ndarray(outputs)
```

Comment:
Do you need to check consistency of the distributed axis between the outputs and the inputs?

Reply:
In practice, it is hard to do from the inputs alone (the axis might change over the course of the operation), but I can check if the expected dist axis matches the dist axis of the outputs!